veusz: init at 3.3.1

laikq opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
A friend of mine uses this program for everyday scientific plotting :)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

laikq committed:
veusz: init at 3.3.1

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,60 @@
+, lib
+}:
+
+let
+  py = python3Packages;
+in
+py.buildPythonApplication rec {
```
```suggestion

python3Packages.buildPythonApplication rec {
```

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,60 @@
+    QT_QPA_PLATFORM=minimal tests/runselftest.py
+  '';
+
+  src = py.fetchPypi {
```
Please put this below version = ... .

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,60 @@
+    "--qt-include-dir=${qtbase.dev}/include"
+  ];
+
+  propagatedBuildInputs = with py; [
```
```suggestion
  propagatedBuildInputs = with python3Packages; [
```

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,60 @@
+
+  installCheckPhase = ''
+    substituteInPlace tests/runselftest.py \
+      --replace "/usr/bin/env python3" "${py.python}/bin/python"
```
```suggestion
      --replace "/usr/bin/env python3" "${python3Packages.python}/bin/python"
```

SuperSandro2000 commented:
Result of `nixpkgs-review pr 93897` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>veusz</li>
  </ul>
</details>

laikq commented:
I implemented your changes and rebased onto master. `nixpkgs-review pr 93897` works without any errors on my x86_64 machine.

SuperSandro2000 commented:
Fails to build on linux:

```
these derivations will be built:
  /nix/store/yqhqq712n4jkzp1n6r6w4clxh4mhwqin-python3.8-astropy-4.2.drv
  /nix/store/xx1zamvhzni1k25j8ydmigy2nhbxav8p-veusz-3.3.drv
building '/nix/store/yqhqq712n4jkzp1n6r6w4clxh4mhwqin-python3.8-astropy-4.2.drv'...
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing setuptools-build-hook
Using setuptoolsBuildPhase
Using setuptoolsShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
unpacking sources
unpacking source archive /nix/store/97p1snwyary26glch3i0i9mkd0m4p7vp-astropy-4.2.tar.gz
source root is astropy-4.2
setting SOURCE_DATE_EPOCH to timestamp 1606289346 of file astropy-4.2/setup.cfg
patching sources
substituteStream(): WARNING: pattern 'auto_use = True' doesn't match anything in file 'setup.cfg'
configuring
no configure script, doing nothing
building
Executing setuptoolsBuildPhase
Traceback (most recent call last):
  File "nix_run_setup", line 8, in <module>
    exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\\r\\n', '\\n'), __file__, 'exec'))
  File "setup.py", line 67, in <module>
    from extension_helpers import get_extensions  # noqa
ModuleNotFoundError: No module named 'extension_helpers'
builder for '/nix/store/yqhqq712n4jkzp1n6r6w4clxh4mhwqin-python3.8-astropy-4.2.drv' failed with exit code 1
cannot build derivation '/nix/store/xx1zamvhzni1k25j8ydmigy2nhbxav8p-veusz-3.3.drv': 1 dependencies couldn't be built
error: build of '/nix/store/xx1zamvhzni1k25j8ydmigy2nhbxav8p-veusz-3.3.drv' failed
```

laikq commented:
Updated veusz to 3.3.1 and removed astropy as a dependency, as it currently fails to build (it's also an optional dependency, so veusz will still work without astropy). I couldn't run `nixpkgs-review` this time (not enough memory on my machine), but `nix-build` worked fine for me.

SuperSandro2000 commented:
Result of `nixpkgs-review pr 93897` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>veusz</li>
  </ul>
</details>

vale981 commented:
@SuperSandro2000 Any blockers here?
:)

SuperSandro2000 commented:
@ofborg eval

SuperSandro2000 commented:
> @SuperSandro2000 Any blockers here?
> :)

The build now fails with ``RuntimeError: Could not find sip executable``. Can you take a look at this if you want the PR to get merged?

laikq commented:
It seems that a change in `sip` caused the error, I added a patch in the package (for some reason, the build script ignores one command line option if the `sip` executable is called `sip5`).

Hmm, can I call ofborg myself?

@ofborg eval

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,71 @@
+  # For some reason, if sip5 is found on the PATH, the option --sip-dir is
+  # ignored in setupPyBuildFlags, see
+  # https://github.com/veusz/veusz/blob/53b99dffa999f2bc41fdc5335d7797ae857c761f/pyqtdistutils.py#L292
+  patchPhase = ''
```
```suggestion
  postPatch = ''
```

laikq replied:
I've added this change now. I guess it is better because one can still define patch files?

SuperSandro2000 replied:
Yes

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,71 @@
+
+  buildInputs = [ qtbase ];
+
+  dontWrapQtApps = true;
+  preFixup = ''
+    wrapQtApp "$out/bin/veusz"
+  '';
```
Why does the auto wrapping not work for us?

laikq replied:
If I comment out these lines, I get the error
```
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```
I'm fairly certain this is because `wrapQtAppsHook` ignores non-ELF executables (and veusz is a `/bin/bash -e` script), as described in the manual [here](https://nixos.org/manual/nixpkgs/stable/#qt-runtime-dependencies).

laikq replied:
I've added a comment in the code to describe why I did this :)

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,73 @@
+  ];
+
+  installCheckPhase = ''
+    substituteInPlace tests/runselftest.py \
+      --replace "/usr/bin/env python3" "${python3Packages.python}/bin/python"
```
```suggestion
```

SuperSandro2000 commented on pkgs/applications/graphics/veusz/default.nix:
```diff
@@ -0,0 +1,73 @@
+  # https://github.com/veusz/veusz/blob/53b99dffa999f2bc41fdc5335d7797ae857c761f/pyqtdistutils.py#L292
+  postPatch = ''
+    substituteInPlace pyqtdistutils.py \
+      --replace "'-I', pyqt5_include_dir," \
+                "'-I', '${python3Packages.pyqt5}/share/sip/PyQt5',"
```
```suggestion
      --replace "'-I', pyqt5_include_dir," "'-I', '${python3Packages.pyqt5}/share/sip/PyQt5',"
    patchShebang tests/runselftest.py
```

laikq commented:
Implemented the changes you suggested!
